### PR TITLE
Assy solver fix

### LINKED ISSUE #767: Need to run Assemble.solve() twice.
From google group discussion: https://groups.google.com/g/cadquery/c/LUeih7iqVrk
Assembly.solve() needs to be run twice otherwise the model creation fails.
Code is not minimal but the error is reproducible.

```python
import cadquery as cq
from cadquery import exporters

class Trough():

    def __init__(self, width, length, height):
        self.width = width
        self.length = length
        self.height = height
        self.lip = 50
        self.thickness = 6
        self.assembly = cq.Assembly()

        self.add_base_to_assembly()
        self.add_sides_to_assembly()
        self.add_back_to_assembly()
        self.add_lips_to_assembly()
        self.assembly.solve() # run once and fail.
        self.assembly.solve() # run twice and all good.

    def make_base_plate(self):
        return cq.Workplane().box(
            self.width + 2 * self.thickness + 2 * self.lip, 
            self.length, 
            self.thickness)

    def make_side_plate(self):
        return cq.Workplane().box(self.length, self.height, self.thickness)

    def make_back_plate(self):
        return cq.Workplane().box(
            self.width + 2 * self.lip + 2 * self.thickness, 
            self.height + self.lip, 
            self.thickness)

    def make_lip_plate(self):
        return cq.Workplane().box(self.length, self.lip, self.thickness)

    def add_base_to_assembly(self):
        self.assembly.add(self.make_base_plate(), name='base_plate')

    def add_sides_to_assembly(self):
        x = self.width / 2 - self.thickness
        translations = (
            ((x, 0, 0), 'side_plate_lhs', '>X'),
            ((-x, 0, 0), 'side_plate_rhs', '<X')
        )
        for coord, name, direction in translations:
            (
                self.assembly
                .add(self.make_side_plate(), name=name)
                .constrain('base_plate',
                           self.make_base_plate().faces('>Z').val().translate(coord), 
                           name, 
                           self.make_side_plate().faces('>Y').val(),
                           'Plane')
                .constrain('base_plate@faces@>Y', '{}@faces@{}'.format(name, direction), 'Axis')
            )
    
    def add_back_to_assembly(self):
        translation = (0, 0, self.height / 2 - self.lip / 2 + self.thickness/2)
        (
            self.assembly
            .add(self.make_back_plate(), name='back_plate')
            .constrain('base_plate',
                       self.make_base_plate().faces('>Y').val().translate(translation), 
                       'back_plate', 
                       self.make_back_plate().faces('>Z').val(),
                       'Plane')
            .constrain('base_plate@faces@>Z', 'back_plate@faces@>Y', 'Axis')
        )

    def add_lips_to_assembly(self):
        translation = (self.width/2, 0, self.height - self.thickness)
        (
            self.assembly
            .add(self.make_lip_plate(), name='lip_lhs')
            .constrain('base_plate@faces@>X', 'lip_lhs@faces@>Y', 'Axis')
            .constrain('base_plate',
                       self.make_base_plate().faces('>Z').val().translate(translation), 
                       'lip_lhs', 
                       self.make_lip_plate().faces('>Z').val(),
                       'Plane')
        )

width = 900
length = 1000
height = 300

mytrough = Trough(width, length, height)
if "show_object" in locals():
    show_object(mytrough.assembly)
```

### PR DESCRIPTION
**Pull request by adam-urbanczyk**
Will resolve #767, #790, #809

### CODE REVIEW

**Comment by marcus7070** [comment]
The docs on the CCSAQ method are an interesting read: https://nlopt.readthedocs.io/en/latest/NLopt_Algorithms/#mma-method-of-moving-asymptotes-and-ccsa

> popular for topology optimization...

Sounds good @adam-urbanczyk!

**Comment by marcus7070** [comment]
Tested on a random assembly with some kind of conflicting constraints. Old `scipy.optimize` solver:
```
Warning: Maximum number of iterations has been exceeded.
         Current function value: 0.000764
         Iterations: 1000
         Function evaluations: 1270
         Gradient evaluations: 1270
```

New nlopt solver:
```
{'cost': 6.476835679796388e-15, 'iters': 1554, 'status': -inf}
```

Assuming iterations and values are comparable between the two techniques, that's possibly an improvement for this one case. Certainly not a regression.

File: cadquery/occ_impl/solver.py [hunk 1]
```diff
@@ -228,34 +226,50 @@ def jac(x):
...
+            "status": opt.last_optimize_result(),
+        }
+
+        return (
+            [
+                Location(self._build_transform(*x[NDOF * i : NDOF * (i + 1)]))
+                for i in range(self.ne)
+            ],
+            result,
+        )
```
Comment (marcus7070) [inline cadquery/occ_impl/solver.py:275]: Just a neater style that still passes black:
```suggestion
        locs = [
            Location(self._build_transform(*x[NDOF * i : NDOF * (i + 1)]))
            for i in range(self.ne)
        ]
        return locs, result
```

**Review by marcus7070 — APPROVED** [review]

**Comment by marcus7070** [comment]
The appveyor check failed on `test_constraint`.

**Comment by adam-urbanczyk** [comment]
Yes, weird

**Review by jmwright — APPROVED** [review]
Looks good, but some comments in cadquery/occ_impl/solver.py would be very welcome.

**Comment by adam-urbanczyk** [comment]
I did run the assy tests 1000 times locally and cannot reproduce the behavior.

**Comment by marcus7070** [comment]
@slobberingant has reported this does resolve #767 and that test failure can not be reproduced, so I'm going to merge.

**Pull request merged**